src/main/java/com/google/sps/DataGraph.java

```diff
@@ -295,8 +322,33 @@ private GraphNode changeNodeToken(GraphNode node, TokenMutation tokenMut) {
     TokenMutation.Type tokenMutType = tokenMut.getType();
     if (tokenMutType == TokenMutation.Type.ADD_TOKEN) {
       tokenList.addAll(tokenNames);
+
+      // Update the map
+      for (String tokenName : tokenNames) {
+        Set<String> nodesWithToken;
+        if (this.tokenMap().containsKey(tokenName)) {
+          nodesWithToken = this.tokenMap().get(tokenName);
+
+        } else {
+          nodesWithToken = new HashSet<>();
+        }
+        nodesWithToken.add(node.name());
+        this.tokenMap().put(tokenName, nodesWithToken);
+      }
     } else if (tokenMutType == TokenMutation.Type.DELETE_TOKEN) {
       tokenList.removeAll(tokenNames);
+      // Update the map
+      for (String tokenName : tokenNames) {
+        if (this.tokenMap().containsKey(tokenName)) {
+          Set<String> nodesWithToken = this.tokenMap().get(tokenName);
+          nodesWithToken.remove(node.name());
+          if (nodesWithToken.size() == 0) { // No more nodes with token
+            this.tokenMap().remove(tokenName);
+          } else {
+            this.tokenMap().put(tokenName, nodesWithToken);
```

Comment:
not sure where to put this but when you filter by token, its not immediately clear which nodes have said token, in my searchToken branch i have some functions that handle highlighting by token and resetting them if you want to look at that

Reply:
passed in the nodes user searched for for changed the color!